File: 프로그래머스/1/172928. 공원 산책/공원 산책.py
def solution(park, routes):
    row = len(park)
    col = len(park[0])
    dirs = {
        "N": (-1,0),
        "S": (1,0),
        "W": (0,-1),
        "E": (0,1),
    }
    
    found = False
    for r in range(row):
        for c in range(col):
            if park[r][c] == "S":
                sr,sc = r,c
                found = True
                break
        if found:
            break
    
    def check_route(a,b,c,d,n):
        for i in range(1,n+1):
            nr,nc=a+c*i,b+d*i
            if 0<=nr<row and 0<=nc<col and park[nr][nc] != "X":
                continue
            else:
                return False
        return True
    
    for route in routes:
        op, n = route.split()
        dr,dc = dirs[op]
        if check_route(sr,sc,dr,dc,int(n)):
            sr,sc = sr+dr*int(n), sc+dc*int(n)
    
    return [sr,sc]

File: 프로그래머스/1/172928. 공원 산책/공원 산책.swift
import Foundation

func solution(_ park:[String], _ routes:[String]) -> [Int] {
    let r = park.count
    let c = park[0].count
    var x = 0, y = 0
    
    var board = park.map {Array($0)}
    var dirs = [
        "E": (0,1), "W": (0,-1), "N": (-1,0), "S": (1,0)
    ]
    
    for i in 0..<r {
        for j in 0..<c {
            if board[i][j] == "S" {
                x = i
                y = j
                break
            }
        }
    }
    
    for route in routes {
        let tmp = Array(route.split(separator: " "))
        let dir = String(tmp[0])
        let dist = Int(tmp[1])!
        
        var nx = x
        var ny = y
        var canMove = true
        
        for _ in 0..<dist {
            nx += dirs[dir]!.0
            ny += dirs[dir]!.1
            
            if 0>nx || nx>=r || 0>ny || ny>=c {
                canMove = false
                break
            }
            
            if(board[nx][ny] == "X") {
                canMove = false
                break
            }
        }
        if canMove {
            x = nx
            y = ny
        }
    }
    
    return [x,y]
}